cores/arduino/stm32/dma.c

```diff
@@ -0,0 +1,315 @@
+  #endif
+  
+  #if defined(STM32F4xx)
+  #ifdef ADC1_BASE
+  if (hadc->Instance == ADC1)
+  {
+    DmaHandleTab[indice].Init.Channel             = DMA_CHANNEL_0;
+  }
+  #endif
+  #ifdef ADC2_BASE
+  if (hadc->Instance == ADC2)
+  {
+    DmaHandleTab[indice].Init.Channel             = DMA_CHANNEL_1;
+  }
+  #endif
+  #ifdef ADC3_BASE
+  if (hadc->Instance == ADC3)
+  {
+    DmaHandleTab[indice].Init.Channel             = DMA_CHANNEL_2;
+  }
+  #endif
+  DmaHandleTab[indice].Init.FIFOMode              = DMA_FIFOMODE_DISABLE;
+  DmaHandleTab[indice].Init.FIFOThreshold         = DMA_FIFO_THRESHOLD_HALFFULL;
+  DmaHandleTab[indice].Init.MemBurst              = DMA_MBURST_SINGLE;
+  DmaHandleTab[indice].Init.PeriphBurst           = DMA_PBURST_SINGLE;
+  #endif //STM32F4xx
+  
+  /* Deinitialize  & Initialize the DMA for new transfer */
+  HAL_DMA_DeInit(&DmaHandleTab[indice]);
+  HAL_DMA_Init(&DmaHandleTab[indice]);
+  
+  /* Associate the DMA handle */
+  __HAL_LINKDMA(hadc, DMA_Handle, DmaHandleTab[indice]);
+
+  /* Set clock as active */
+  channelNumber = pinmap_find_dmaChannelNumber(hadc->Instance, DmaMap_DMA);
+  dmaNumber     = pinmap_find_dma(hadc->Instance, DmaMap_DMA);
+  dmaTab[dmaNumber-1][channelNumber] = true;
+
+  /* NVIC configuration for DMA Input data interrupt */
+  irqHandler = pinmap_find_dmaIrqHandler(hadc->Instance, DmaMap_DMA);
+  HAL_NVIC_SetPriority(irqHandler, 1, 0);
+  HAL_NVIC_EnableIRQ(irqHandler);
+}
+
+/**
+  * @brief  This function will deinit the dma and channel corresponding 
+  *         to the ADC peripheral
+  * @param  hadc : the ADC peripheral to use
+  * @retval none
+  */
+void dma_deinit_adc(ADC_HandleTypeDef *hadc)
+{
+  IRQn_Type           irqHandler;
+  int8_t              channelNumber;
+  int8_t              dmaNumber;
+  
+  irqHandler = pinmap_find_dmaIrqHandler(hadc->Instance, DmaMap_DMA);
+  HAL_NVIC_DisableIRQ(irqHandler);
+  HAL_DMA_DeInit(hadc->DMA_Handle);
+  
+  // Disable the clock only if not another channel of the dma is use
+  channelNumber = pinmap_find_dmaChannelNumber(hadc->Instance, DmaMap_DMA);
+  dmaNumber     = pinmap_find_dma(hadc->Instance, DmaMap_DMA);
+  dmaTab[dmaNumber-1][channelNumber] = false;
+  
+ for (int i=0; i<CHANNEL_NUMBER; i++)
+  {
+    if (dmaTab[dmaNumber][i] == true)
+    {
+      break;
+    }
+    if (i == CHANNEL_NUMBER-1)
+    {
+      // All dma channel are inactive
+      if (dmaNumber == 1)
+      {
+      #ifdef __HAL_RCC_DMA1_CLK_ENABLE
+        __HAL_RCC_DMA1_CLK_DISABLE();
+      #endif
+      }
+      else if (dmaNumber == 2)
+      {
+        #ifdef __HAL_RCC_DMA2_CLK_ENABLE
+        __HAL_RCC_DMA2_CLK_DISABLE();
+        #endif
+      }
+    }
+  }
+}
+
+
+/**
+  * @brief  DMA IRQ Handle 
+  * @param  None
+  * @retval None
+  */
+
+#ifdef ADC1_BASE
+void DMA_ADC1_IRQHandler1(void)
+{
+  #ifdef STM32F4xx
+  if (AdcHandleTab[0].DMA_Handle != NULL)
+  {
+    HAL_DMA_IRQHandler(AdcHandleTab[0].DMA_Handle); // ADC1
+  }
+  else
```

Comment:
could be 
elsif (AdcHandleTab[0].DMA_Handle != NULL)

Reply:
I don't really understand what you mean. If and elseif can't have the same condition. I can put a condition on the elseif if that's what you want.